pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy

```diff
@@ -308,21 +308,57 @@ class Utils {
         }
     }
 
+    static void markStageWithTag(String stageName, String stepContextFlowNodeId, String tagName, String tagValue) {
+        CpsThread thread = CpsThread.current()
+        FlowExecution execution = thread.execution
+
+        def stepContextFlowNode = execution.getNode(stepContextFlowNodeId)
+
+        List<FlowNode> nodes = []
+
+        FlowNode stage = (FlowNode) stepContextFlowNode.iterateEnclosingBlocks().find { blockStartNode ->
+            CommonUtils.isStageWithOptionalName(stageName).test(blockStartNode)
+        }
+
+        if (stage != null) {
+            nodes.add(stage)
+
+            // Additional check needed to get the possible enclosing parallel branch for a nested stage.
+            Filterator<FlowNode> filtered = FlowScanningUtils.fetchEnclosingBlocks(stage)
+                    .filter(isParallelBranchFlowNode(stageName))
+
+            filtered.each { f ->
+                if (f != null) {
+                    nodes.add(f)
+                }
+            }
+        }
+
+        nodes.each { currentNode ->
+            addTagToFlowNode(currentNode, tagName, tagValue)
+        }
```

Comment:
`FlowScanningUtils.fetchEnclosingBlocks(stage)` should just return a subset of `stepContextFlowNode.iterateEnclosingBlocks()` (and IDK why the method even exists in the first place). I think you should be able to simplify this:

```suggestion
        for (FlowNode node : stepContextFlowNode.iterateEnclosingBlocks()) {
            if (CommonUtils.isStageWithOptionalName(stageName).test(node) || isParallelBranchFlowNode(stageName).apply(node)) {
                addTagToFlowNode(node, tagName, tagValue)
            } 
        }
```

Really though IIUC, the problem has to do with [`CommonUtils.isStageWithOptionalName`](https://github.com/jenkinsci/pipeline-model-definition-plugin/blob/b364814683740de3e3af94e521df8d5f9ca8c544/pipeline-model-extensions/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/CommonUtils.java#L52) returning true both for the stage step itself as the node for the parallel branch with the same name, so maybe `isParallelBranchFlowNode` is redundant once you switch to iterating over all of the enclosing blocks. Also it might be best to stop iterating once you hit two matches in case the Pipeline is really complicated.

Visualizations are probably only targeting one or the other of those nodes, so I guess you were targeting the parallel branch node rather than the stage, or something like that? I recall having to do something weird in this case too for my visualization, maybe checking both the parallel branch and the stage step for the skipped tag.

Reply:
yes the parallel branch node is the one that is visualised from what I can tell.

The tests in this plugin expect both the branch and the stage to have the skipped tag on it.

Reply:
(edge case I hit while testing this which includes some JSON to show what is used if you are interested https://github.com/jenkinsci/pipeline-graph-view-plugin/issues/358)